### ISSUE TITLE
Using CanDrag as bound dependency property causes UNCACTHEABLE crash in d3d11

### ISSUE THREAD

**Issue by jtorjo** [microsoft/microsoft-ui-xaml#2056]
**Describe the bug**

I have an ItemsControl, with items placed from left to right.

Around edges, my code will catch that and allow the item to be resized. Otherwise, items can be dragged around (so user can move the items where he wants to).

This crashes when I resize an item, stop, then resize it again (when moving mouse very fast).

**Steps to reproduce the bug**

Side note: This was **insanely** hard to catch.

	<ItemsControl ItemsSource="{Binding medias}">
		<ItemsControl.ItemsPanel>
			<ItemsPanelTemplate>
				<StackPanel Orientation="Horizontal" />
			</ItemsPanelTemplate>
		</ItemsControl.ItemsPanel>
		<ItemsControl.ItemTemplate>
			<DataTemplate>
				<Grid Width="{Binding width_seconds}" 
					  CanDrag="{Binding can_drag}" AllowDrop="True" 
                                          PointerPressed="pic_down"  ... >
						...
				</Grid>
			</DataTemplate>
		</ItemsControl.ItemTemplate>
	</ItemsControl>

I'm updating `can_drag` on `pic_down` event.

For some strange reason, setting `can_drag` to false while starting to resize (which Windows can also interpret as the start of a drag), will end up crashing the application. And once again, it will crash like a boss - you can't do anything about it.

![image](https://user-images.githubusercontent.com/13891601/75800568-71869b00-5d82-11ea-8af5-9c0ce872e57f.png)

And if you think you can actually attach a debugger, think again:

![image](https://user-images.githubusercontent.com/13891601/75800646-94b14a80-5d82-11ea-8103-fde7644eb3d6.png)

A bit more info

![image](https://user-images.githubusercontent.com/13891601/75800854-de9a3080-5d82-11ea-967e-2a52f0c27124.png)

and 

![image](https://user-images.githubusercontent.com/13891601/75800880-e9ed5c00-5d82-11ea-824e-d9c67832a412.png)

You can have a coffee, a glass of wine, two, it will never end.

**Solution**

Don't use CanDrag as a bound property. The only place I could safely set CanDrag and not have the app crash is on `PointerDown` event.

**A bit more info**

**This should NEVER EVER EVER EVER happen.**

It should never matter what my code does. I give you bad input, you (Microsoft) should throw an exception. But **you shoud never ever crash my app**.

| Windows 10 version                  | Saw the problem? |
| :--------------------------------- | :-------------------- |
| November 2019 Update (18363)  | Yes   |

| Device form factor | Saw the problem? |
| :-------------------- | :------------------- |
| Desktop                 | Yes |

**Additional context**

**Comment by ranjeshj** [comment]
@jtorjo Thanks for reporting this crash. Crashes in native code (access violations for example) do not get bubbled up to the managed layer to be able to provide a .net exception. You can enable native exceptions in visual studio to see the entire stack (native and managed) when the crash happens to see exactly where the failure happened. Unfortunately screenshots/error reporting don't really help understand the issue. A crash memory dump or call stack at the point of failure would be much more actionable.

![image](https://user-images.githubusercontent.com/28935693/75806704-095f9700-5d39-11ea-9908-f1e4e35500df.png)

I am having trouble reproducing the crash from the snippet you have above. If you can share a standalone repro app, that would greatly help track this crash down. Crashes are not intentional. If input is incorrect, we throw exceptions, so this is a bug.

**Comment by jtorjo** [comment]
@ranjeshj 

I've enabled native exceptions

![image](https://user-images.githubusercontent.com/13891601/75815450-195c9280-5d9c-11ea-9fa2-f5c1cb71259c.png)

![image](https://user-images.githubusercontent.com/13891601/75815663-78baa280-5d9c-11ea-911e-d764e67a5cfd.png)

The same thing happens.

I can't share a standalone app - there's a lot of code involved.

**Comment by ranjeshj** [comment]
Would it be possible for you to capture a crash dump and share that ? Instructions can be found here - https://docs.microsoft.com/en-us/windows/win32/wer/collecting-user-mode-dumps . Use the Full dump type.

**Comment by jtorjo** [comment]
@ranjeshj 

Here it is - https://phot-awe.com/ScriptedSandbox64.exe.8800.dmp.zip - please let me know after you've downloaded it, so I can delete it.

Also, please let me know if it's ok.

**Comment by ranjeshj** [comment]
Thanks. I got it, so you can remove the dump from your share. This seems to be a crash in d3d11. @codendone Can you please route this bug ?


```
 # Child-SP          RetAddr           Call Site
00 00000013`b20fd970 00007ffc`83b28aa9 KERNELBASE!RaiseFailFastException(struct _EXCEPTION_RECORD * pExceptionRecord = 0x00000013`b20fdfc0, struct _CONTEXT * pContextRecord = 0x00000013`b20fda50, unsigned long dwFlags = 0)+0xaf [minkernel\kernelbase\xcpt.c @ 1133] 
01 00000013`b20fdf40 00007ffc`83b28b74 d3d11!wil::details::WilDynamicLoadRaiseFailFastException(struct _EXCEPTION_RECORD * er = 0x00000013`b20fdfc0, struct _CONTEXT * cr = 0x00000000`00000000, unsigned long flags = 0)+0x49 [internal\sdk\inc\wil\opensource\wil\result_macros.h @ 1694] 
02 00000013`b20fdf70 00007ffc`83b28b53 d3d11!wil::details::WilRaiseFailFastException(struct _EXCEPTION_RECORD * er = 0x00000013`b20fda50, unsigned long flags = 0)+0x18 [internal\sdk\inc\wil\opensource\wil\result_macros.h @ 1637] 
03 00000013`b20fdfa0 00007ffc`83b28179 d3d11!wil::details::WilFailFast(struct wil::FailureInfo * failure = 0x00000013`b20fe0f0)+0x93 [internal\sdk\inc\wil\opensource\wil\result_macros.h @ 3317] 
04 00000013`b20fe070 00007ffc`83b282e9 d3d11!wil::details::ReportFailure(void * callerReturnAddress = <Value unavailable error>, unsigned int lineNumber = <Value unavailable error>, char * fileName = <Value unavailable error>, char * functionName = <Value unavailable error>, char * code = 0x00000000`00000000 "", void * returnAddress = 0x00007ffc`83b16eb1, wil::FailureType type = FailFast (0n3), HRESULT hr = 0x8000ffff)+0xe9 [internal\sdk\inc\wil\opensource\wil\result_macros.h @ 3356] 
05 00000013`b20ff5b0 00007ffc`83b27140 d3d11!wil::details::ReportFailure_Hr(void * callerReturnAddress = <Value unavailable error>, unsigned int lineNumber = <Value unavailable error>, char * fileName = <Value unavailable error>, char * functionName = <Value unavailable error>, char * code = 0x00000000`00000000 "", void * returnAddress = 0x00007ffc`83b16eb1, wil::FailureType type = FailFast (0n3), HRESULT hr = 0x8000ffff)+0x39 [internal\sdk\inc\wil\opensource\wil\result_macros.h @ 3414] 
06 00000013`b20ff610 00007ffc`83b16eb1 d3d11!wil::details::in1diag3::FailFast_Unexpected(void * callerReturnAddress = <Value unavailable error>)+0x38 [internal\sdk\inc\wil\opensource\wil\result_macros.h @ 4695] 
07 00000013`b20ff660 00007ffc`83aa7012 d3d11!wil::mutex_t<wil::details::unique_storage<wil::details::resource_policy<void *,void +0x6debd [internal\sdk\inc\wil\opensource\wil\resource.h @ 2838] 
08 (Inline Function) --------`-------- d3d11!wil::details_abi::ProcessLocalStorageData<wil::details_abi::FeatureStateData>::Release(void)+0xe [internal\sdk\inc\wil\opensource\wil\result.h @ 243] 
09 (Inline Function) --------`-------- d3d11!wil::details_abi::ProcessLocalStorage<wil::details_abi::FeatureStateData>::{dtor}(void)+0x21 [internal\sdk\inc\wil\opensource\wil\result.h @ 334] 
0a (Inline Function) --------`-------- d3d11!wil::details::FeatureStateManager::ProcessShutdown(void)+0x21 [internal\sdk\inc\wil\staging.h @ 4072] 
0b 00000013`b20ff690 00007ffc`83af59c2 d3d11!wil::manually_managed_shutdown_aware_object<wil::details::FeatureStateManager>::destroy(void)+0x46 [internal\sdk\inc\wil\opensource\wil\result_macros.h @ 2055] 
0c (Inline Function) --------`-------- d3d11!wil::shutdown_aware_object<wil::details::FeatureStateManager>::{dtor}(void)+0x5 [internal\sdk\inc\wil\opensource\wil\result_macros.h @ 2087] 
0d 00000013`b20ff6d0 00007ffc`83ae931f d3d11!wil::details::`dynamic atexit destructor for 'g_featureStateManager''(void)+0x12
0e 00000013`b20ff710 00007ffc`83ae96ac d3d11!_CRT_INIT(void * hDllHandle = <Value unavailable error>, unsigned long dwReason = <Value unavailable error>, void * lpreserved = <Value unavailable error>)+0xc3 [minkernel\crts\crtw32\dllstuff\crtdll.c @ 506] 
0f 00000013`b20ff750 00007ffc`89f850a1 d3d11!__DllMainCRTStartup(void * hDllHandle = 0x00007ffc`83a70000, unsigned long dwReason = 0, void * lpreserved = 0x00000000`00000001)+0x1c8 [minkernel\crts\crtw32\dllstuff\crtdll.c @ 658] 
10 00000013`b20ff8c0 00007ffc`89fcab02 ntdll!LdrpCallInitRoutine(<function> * InitRoutine = 0x00007ffc`83ae94a0, void * DllHandle = 0x00007ffc`83a70000, unsigned long Reason = 0, struct _CONTEXT * Context = 0x00000000`00000001)+0x65 [minkernel\ntdll\ldr.c @ 212] 
11 00000013`b20ff930 00007ffc`89fca9ad ntdll!LdrShutdownProcess(void)+0x132 [minkernel\ntdll\ldrinit.c @ 6224] 
12 00000013`b20ffa30 00007ffc`88a7cd8a ntdll!RtlExitUserProcess(long ExitStatus = 0n0)+0xad [minkernel\ntdll\rtlstrt.c @ 1571] 
13 00000013`b20ffa60 00007ff7`d848c6c4 kernel32!FatalExit(int ExitCode = <Value unavailable error>)+0xa [base\win32\client\process.c @ 2421] 
14 00000013`b20ffa90 00007ff7`d848c660 ScriptedSandbox64!exit_or_terminate_process(unsigned int return_code = 0)+0x48 [d:\th\minkernel\crts\ucrt\src\appcrt\startup\exit.cpp @ 129] 
15 00000013`b20ffac0 00007ff7`d848567c ScriptedSandbox64!common_exit(int return_code = 0n0, _crt_exit_cleanup_mode cleanup_mode = _crt_exit_full_cleanup (0n0), _crt_exit_return_mode return_mode = _crt_exit_terminate_process (0n0))+0x150 [d:\th\minkernel\crts\ucrt\src\appcrt\startup\exit.cpp @ 269] 
16 00000013`b20ffaf0 00007ffc`88a77bd4 ScriptedSandbox64!__scrt_common_main_seh(void)+0x168 [d:\agent\_work\2\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl @ 295] 
17 00000013`b20ffb30 00007ffc`89fcced1 kernel32!BaseThreadInitThunk(unsigned long RunProcessInit = <Value unavailable error>, <function> * StartAddress = <Value unavailable error>, void * Argument = <Value unavailable error>)+0x14 [base\win32\client\thread.c @ 64] 
18 00000013`b20ffb60 00000000`00000000 ntdll!RtlUserThreadStart(<function> * StartAddress = 0x00000000`00000000, void * Argument = 0x00000000`00000000)+0x21 [minkernel\ntdll\rtlstrt.c @ 1153] 
```

**Comment by jtorjo** [comment]
> Thanks. I got it, so you can remove the dump from your share.

@ranjeshj Thanks, removed

**Comment by codendone** [comment]
@jtorjo I think that dump is not from the right process. This ScriptedSandbox64.exe crash might be a side-effect of your app crashing. Please try turning off Tools | Options | Debugging | "Enable Diagnostic Tools while debugging" in Visual Studio and then run your scenario. Hopefully that will cause it to break in for the crash in your app so you can grab a dump of that.

**Comment by jtorjo** [comment]
@codendone https://phot-awe.com/dumps.zip - for some strange reason, two files were created. Archived both. Please let me know after you've downloaded it so I can erase it.

**Comment by codendone** [comment]
Thanks, @jtorjo!

It turns out this is a dup of #1540, which is fixed for the next version of Windows (via internal bug 24011776), and will also be fixed in WinUI 3.0.

The crash stack:
```
Windows_UI_Xaml!DirectUI::AutomaticDragHelper::HandlePointerMovedEventArgs+0x9a [onecoreuap\windows\dxaml\xcp\dxaml\lib\automaticdraghelper.cpp @ 274]
Windows_UI_Xaml!DirectUI::UIElement::OnPointerMoved+0x28 [onecoreuap\windows\dxaml\xcp\dxaml\lib\uielement_partial_dragdrop.cpp @ 232]
Windows_UI_Xaml!DirectUI::ClassMemberEventHandler<DirectUI::UIElement,Windows::UI::Xaml::IUIElement,Windows::UI::Xaml::Input::IPointerEventHandler,IInspectable,Windows::UI::Xaml::Input::IPointerRoutedEventArgs>::Invoke+0x94 [onecoreuap\windows\dxaml\xcp\dxaml\lib\joltclasses.h @ 663]
Windows_UI_Xaml!DirectUI::CRoutedEventSourceBase<DirectUI::IUntypedEventSource,Windows::UI::Xaml::Input::IPointerEventHandler,IInspectable,Windows::UI::Xaml::Input::IPointerRoutedEventArgs>::Raise+0x197 [onecoreuap\windows\dxaml\xcp\dxaml\lib\joltclasses.h @ 1093]
Windows_UI_Xaml!DirectUI::CRoutedEventSourceBase<DirectUI::IUntypedEventSource,Windows::UI::Xaml::Input::IPointerEventHandler,IInspectable,Windows::UI::Xaml::Input::IPointerRoutedEventArgs>::UntypedRaise+0x9a [onecoreuap\windows\dxaml\xcp\dxaml\lib\joltclasses.h @ 995]
Windows_UI_Xaml!DirectUI::DependencyObject::FireEvent+0x102 [onecoreuap\windows\dxaml\xcp\dxaml\lib\dependencyobject.cpp @ 3475]
Windows_UI_Xaml!DirectUI::DXamlCore::FireEvent+0x2fe [onecoreuap\windows\dxaml\xcp\dxaml\lib\dxamlcore.cpp @ 1962]
Windows_UI_Xaml!AgCoreCallbacks::FireEvent+0x22 [onecoreuap\windows\dxaml\xcp\dxaml\lib\fxcallbacks.cpp @ 91]
Windows_UI_Xaml!CFxCallbacks::JoltHelper_FireEvent+0x22 [onecoreuap\windows\dxaml\xcp\dxaml\lib\fxcallbacks.cpp @ 1035]
Windows_UI_Xaml!CCoreServices::CLR_FireEvent+0x394 [onecoreuap\windows\dxaml\xcp\core\dll\xcpcore.cpp @ 3224]
Windows_UI_Xaml!CommonBrowserHost::CLR_FireEvent+0x1a [onecoreuap\windows\dxaml\xcp\control\common\shared\commonbrowserhost.hpp @ 771]
Windows_UI_Xaml!CControlBase::ScriptCallback+0x11e [onecoreuap\windows\dxaml\xcp\control\common\shared\controlbase.cpp @ 267]
Windows_UI_Xaml!CXcpDispatcher::OnScriptCallback+0x112 [onecoreuap\windows\dxaml\xcp\win\shared\xcpwindow.cpp @ 1336]
Windows_UI_Xaml!CXcpDispatcher::OnWindowMessage+0x4f [onecoreuap\windows\dxaml\xcp\win\shared\xcpwindow.cpp @ 1187]
Windows_UI_Xaml!CXcpDispatcher::ProcessMessage+0x8d [onecoreuap\windows\dxaml\xcp\win\shared\xcpwindow.cpp @ 909]
Windows_UI_Xaml!CXcpDispatcher::WindowProc+0x4d [onecoreuap\windows\dxaml\xcp\win\shared\xcpwindow.cpp @ 837]
user32!_InternalCallWinProc+0x2b [internal\onecoreuapwindows\private\lib\i386\Core\ntuser\client\callproc.asm @ 116]
user32!InternalCallWinProc+0x1a [internal\onecoreuapwindows\private\inc\core\ntuser\client\callproc.h @ 87]
user32!UserCallWinProcCheckWow+0x3ac [windows\core\ntuser\client\clmsg.cxx @ 279]
user32!DispatchClientMessage+0xea [windows\core\ntuser\client\wow6432\objfre\i386\client.cxx @ 3442]
user32!__fnDWORD+0x3f [internal\onecoreuapwindows\inc\ntuser\inc\ntcb.h @ 1250]
ntdll!KiUserCallbackDispatcher+0x4d [minkernel\ntos\rtl\i386\userdisp.asm @ 524]
win32u!NtUserMessageCall+0xc [windows\core\umode\wow6432\objfre\i386\usrstubs.asm @ 185]
user32!SendMessageWorker+0x3c9 [windows\core\ntuser\client\clmsg.cxx @ 890]
user32!SendMessageW+0x123 [internal\onecoreuapwindows\private\inc\core\ntuser\client\cltxt.h @ 906]
Windows_UI_Xaml!CXcpDispatcher::SendMessageW+0x10 [onecoreuap\windows\dxaml\xcp\win\shared\xcpwindow.cpp @ 807]
Windows_UI_Xaml!CXcpBrowserHost::SyncScriptCallbackRequest+0xea [onecoreuap\windows\dxaml\xcp\host\win\browserdesktop\winbrowserhost.cpp @ 1048]
Windows_UI_Xaml!CEventManager::RaiseHelper+0x194 [onecoreuap\windows\dxaml\xcp\core\dll\eventmgr.cpp @ 1068]
Windows_UI_Xaml!CEventManager::Raise+0x292 [onecoreuap\windows\dxaml\xcp\core\dll\eventmgr.cpp @ 902]
Windows_UI_Xaml!CEventManager::RaiseRoutedEventBubbling+0xa6 [onecoreuap\windows\dxaml\xcp\core\dll\eventmgr.cpp @ 1334]
Windows_UI_Xaml!CEventManager::RaiseRoutedEvent+0x2e [onecoreuap\windows\dxaml\xcp\core\dll\eventmgr.cpp @ 1266]
Windows_UI_Xaml!ContentRootInput::PointerInputProcessor::ProcessPointerInput+0xa72 [onecoreuap\windows\dxaml\xcp\components\contentroot\pointerinputprocessor.cpp @ 589]
Windows_UI_Xaml!CInputServices::ProcessInput+0x240eaf [onecoreuap\windows\dxaml\xcp\core\input\inputservices.cpp @ 887]
Windows_UI_Xaml!CCoreServices::ProcessInput+0x47 [onecoreuap\windows\dxaml\xcp\core\dll\xcpcore.cpp @ 991]
Windows_UI_Xaml!CXcpBrowserHost::HandleInputMessage+0x2a2 [onecoreuap\windows\dxaml\xcp\host\win\browserdesktop\winbrowserhost.cpp @ 1403]
Windows_UI_Xaml!CJupiterControl::HandlePointerMessage+0xc4 [onecoreuap\windows\dxaml\xcp\dxaml\lib\jupitercontrol.cpp @ 750]
Windows_UI_Xaml!CJupiterControl::HandleWindowMessage+0x240e96 [onecoreuap\windows\dxaml\xcp\dxaml\lib\jupitercontrol.cpp @ 326]
Windows_UI_Xaml!CJupiterWindow::CoreWindowSubclassProc+0x164 [onecoreuap\windows\dxaml\xcp\dxaml\lib\jupiterwindow.cpp @ 1084]
Windows_UI_Xaml!CJupiterWindow::StaticCoreWindowSubclassProc+0x27 [onecoreuap\windows\dxaml\xcp\dxaml\lib\jupiterwindow.cpp @ 576]
user32!_InternalCallWinProc+0x2b [internal\onecoreuapwindows\private\lib\i386\Core\ntuser\client\callproc.asm @ 116]
user32!InternalCallWinProc+0x1a [internal\onecoreuapwindows\private\inc\core\ntuser\client\callproc.h @ 87]
user32!UserCallWinProcCheckWow+0x3ac [windows\core\ntuser\client\clmsg.cxx @ 279]
user32!DispatchMessageWorker+0x20e [windows\core\ntuser\client\clmsg.cxx @ 3142]
user32!DispatchMessageW+0x10 [internal\onecoreuapwindows\private\inc\core\ntuser\client\cltxt.h @ 1101]
Windows_UI!Windows::UI::Core::CDispatcher::ProcessMessage+0x105 [onecoreuap\windows\advcore\winrt\onecoreiwindow\corewindow\common\dispatcher.cpp @ 327]
Windows_UI!Windows::UI::Core::CDispatcher::WaitAndProcessMessagesInternal+0x9d [onecoreuap\windows\advcore\winrt\onecoreiwindow\corewindow\common\dispatcher.cpp @ 1953]
Windows_UI!Windows::UI::Core::CDispatcher::ProcessEvents+0x65 [onecoreuap\windows\advcore\winrt\onecoreiwindow\corewindow\common\dispatcher.cpp @ 581]
Windows_UI_Xaml!CJupiterWindow::RunCoreWindowMessageLoop+0x50 [onecoreuap\windows\dxaml\xcp\dxaml\lib\jupiterwindow.cpp @ 1233]
Windows_UI_Xaml!CJupiterControl::RunMessageLoop+0x67 [onecoreuap\windows\dxaml\xcp\dxaml\lib\jupitercontrol.cpp @ 1065]
Windows_UI_Xaml!DirectUI::DXamlCore::RunMessageLoop+0x75 [onecoreuap\windows\dxaml\xcp\dxaml\lib\dxamlcore.cpp @ 2453]
Windows_UI_Xaml!DirectUI::FrameworkView::Run+0x8c [onecoreuap\windows\dxaml\xcp\dxaml\lib\frameworkview_partial.cpp @ 101]
twinapi_appcore!Windows::ApplicationModel::Core::CoreApplicationView::Run+0x32 [onecoreuap\shell\coreapplication\application\lib\coreapplicationview.cpp @ 561]
twinapi_appcore!Windows::ApplicationModel::Core::CoreApplicationViewAgileContainer::RuntimeClassInitialize::__l2::<lambda_a740d69e4550265529206848c8e72a27>::operator()+0x7f [onecoreuap\shell\coreapplication\application\lib\coreapplicationviewagilecontainer.cpp @ 1155]
twinapi_appcore!<lambda_a740d69e4550265529206848c8e72a27>::<lambda_invoker_stdcall>+0x8b [onecoreuap\shell\coreapplication\application\lib\coreapplicationviewagilecontainer.cpp @ 1184]
SHCore!_WrapperThreadProc+0xe0 [onecore\shell\shcore\libs\thread\thread.cpp @ 321]
kernel32!BaseThreadInitThunk+0x19 [base\win32\client\thread.c @ 64]
ntdll!__RtlUserThreadStart+0x2f [minkernel\ntdll\rtlstrt.c @ 1153]
ntdll!_RtlUserThreadStart+0x1b [minkernel\ntdll\rtlstrt.c @ 1070]
```

**Comment by jtorjo** [comment]
@codendone Awesome, thanks!
